src/librbd/internal.cc

```diff
@@ -1706,11 +1706,17 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
 
       C_SaferCond lock_ctx;
       ictx->exclusive_lock->request_lock(&lock_ctx);
+
+      // don't block holding lock since refresh might be required
+      ictx->owner_lock.put_read();
```

Comment:
@dillaman May be it makes sense then to acquire lock before block_writes and block_requests? (not sure how important it is, is there a chance something will unblock while we are not holding the lock?)

Also, don't we need the same when we are releasing the exclusive lock?

Another thing I have noticed, not related to your change, that after acquiring the lock we don't check for acquired_lock value when returning due to a error below. This looks like a potential lock leakage to me?